### ISSUE TITLE
✨ New factory functions with configurable regular expression for `EmailAddress`

### ISSUE THREAD

**Issue by LVMVRQUXL** [kotools/types#388]
## 📝 Description

We would like to introduce the following **experimental** factory functions with configurable regular expression for the `EmailAddress` type:

```kotlin
fun EmailAddress.Companion.create(
    text: String,
    configuration: EmailAddress.RegexConfiguration.() -> Unit
): EmailAddress

fun EmailAddress.Companion.createOrNull(
    text: String,
    configuration: EmailAddress.RegexConfiguration.() -> Unit
): EmailAddress?

interface EmailAddress.RegexConfiguration {
    var localPart: Regex
    var domain: Regex
}
```

Here's an example of calling the `create(String, EmailAddress.RegexConfiguration.() -> Unit)` function from Kotlin code:

```kotlin
val address: EmailAddress = EmailAddress.create("contact@kotools.org") {
    localPart = Regex("[a-z]+") // the local part now only accepts 1 or more lowercase letters
    domain = Regex("[a-z]+\.[a-z]+") // the domain now only accepts lowercase letters
    // The regular expression used for validating the text input now looks like this: ^[a-z]+@[a-z]+\.[a-z]+$
}
println(address) // contact@kotools.org
```

## 🔗 Dependencies

This issue is blocked by the following ones:
- [x] #339 
- [x] #394 

## ✅ Checklist

- [ ] Add the `create(String, EmailAddress.RegexConfiguration.() -> Unit)` function, [test] its behavior with Kotlin and its compatibility with Java, update the public [API binaries] and update the [unreleased changelog].
- [ ] Add the `createOrNull(String, EmailAddress.RegexConfiguration.() -> Unit)` function, [test] its behavior with Kotlin and its compatibility with Java, update the public [API binaries] and update the [unreleased changelog].
- [ ] Close this issue as completed and update tracking ones if present.

[api binaries]: https://github.com/kotools/types/blob/main/CONTRIBUTING.md#checking-the-api-binaries
[test]: https://github.com/kotools/types/blob/main/CONTRIBUTING.md#running-tests
[unreleased changelog]: https://github.com/kotools/types/blob/main/CHANGELOG.md#unreleased

**Comment by LVMVRQUXL** [comment]
The DSL for setting the `domain` property should always include the dot (`.`) between labels.

Here's an example for configuring the domain name's regular expression from Kotlin code:

```kotlin
EmailAddress.create("contact@kotools.org") {
    domain {
        topLevel = Regex("[a-z]+")
        secondLevel = firstLabel
    }
}
```

**Comment by LVMVRQUXL** [comment]
We should take more time to design this solution according to the validation of the regular expression received from users.

**Comment by LVMVRQUXL** [comment]
Another implementation of the `EmailAddress` type will be available after resolving #635.
